[Bug]: `pip install chgnet` failed

### Version

v0.3.7

### Which OS(es) are you using?

- [X] MacOS (M3 Chip, MacOS 14.5)
- [ ] Windows
- [x] Linux (Ubuntu 22.04, AMD EPYC 7763 CPU)

### What happened?

`pip install chgnet` seems to fail for the latest `v0.3.7` release, which also seems to impact the `pymatgen` CI: https://github.com/materialsproject/pymatgen/actions/runs/9350037332/job/25732649581?pr=3822 

### Code snippet

```python
pip install chgnet
```


### Log output

```shell
Building wheels for collected packages: chgnet, nvidia-ml-py3
  Building wheel for chgnet (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Building wheel for chgnet (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [60 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.macosx-11.1-arm64-cpython-311
      creating build/lib.macosx-11.1-arm64-cpython-311/chgnet
      copying chgnet/__init__.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet
      creating build/lib.macosx-11.1-arm64-cpython-311/chgnet/graph
      copying chgnet/graph/graph.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/graph
      copying chgnet/graph/converter.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/graph
      copying chgnet/graph/__init__.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/graph
      copying chgnet/graph/crystalgraph.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/graph
      creating build/lib.macosx-11.1-arm64-cpython-311/chgnet/utils
      copying chgnet/utils/__init__.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/utils
      copying chgnet/utils/vasp_utils.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/utils
      copying chgnet/utils/common_utils.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/utils
      creating build/lib.macosx-11.1-arm64-cpython-311/chgnet/model
      copying chgnet/model/basis.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/model
      copying chgnet/model/functions.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/model
      copying chgnet/model/__init__.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/model
      copying chgnet/model/encoders.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/model
      copying chgnet/model/model.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/model
      copying chgnet/model/composition_model.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/model
      copying chgnet/model/layers.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/model
      copying chgnet/model/dynamics.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/model
      creating build/lib.macosx-11.1-arm64-cpython-311/chgnet/data
      copying chgnet/data/__init__.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/data
      copying chgnet/data/dataset.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/data
      creating build/lib.macosx-11.1-arm64-cpython-311/chgnet/trainer
      copying chgnet/trainer/__init__.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/trainer
      copying chgnet/trainer/trainer.py -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/trainer
      running egg_info
      writing chgnet.egg-info/PKG-INFO
      writing dependency_links to chgnet.egg-info/dependency_links.txt
      writing requirements to chgnet.egg-info/requires.txt
      writing top-level names to chgnet.egg-info/top_level.txt
      reading manifest file 'chgnet.egg-info/SOURCES.txt'
      adding license file 'LICENSE'
      writing manifest file 'chgnet.egg-info/SOURCES.txt'
      copying chgnet/graph/cygraph.pyx -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/graph
      creating build/lib.macosx-11.1-arm64-cpython-311/chgnet/pretrained
      creating build/lib.macosx-11.1-arm64-cpython-311/chgnet/pretrained/0.2.0
      copying chgnet/pretrained/0.2.0/README.md -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/pretrained/0.2.0
      copying chgnet/pretrained/0.2.0/chgnet_0.2.0_e30f77s348m32.pth.tar -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/pretrained/0.2.0
      creating build/lib.macosx-11.1-arm64-cpython-311/chgnet/pretrained/0.3.0
      copying chgnet/pretrained/0.3.0/README.md -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/pretrained/0.3.0
      copying chgnet/pretrained/0.3.0/chgnet_0.3.0_e29f68s314m37.pth.tar -> build/lib.macosx-11.1-arm64-cpython-311/chgnet/pretrained/0.3.0
      running build_ext
      Compiling chgnet/graph/cygraph.pyx because it changed.
      [1/1] Cythonizing chgnet/graph/cygraph.pyx
      building 'chgnet.graph.cygraph' extension
      creating build/temp.macosx-11.1-arm64-cpython-311
      creating build/temp.macosx-11.1-arm64-cpython-311/chgnet
      creating build/temp.macosx-11.1-arm64-cpython-311/chgnet/graph
      clang -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /opt/anaconda3/include -arch arm64 -fPIC -O2 -isystem /opt/anaconda3/include -arch arm64 -I/opt/anaconda3/include/python3.11 -c chgnet/graph/cygraph.c -o build/temp.macosx-11.1-arm64-cpython-311/chgnet/graph/cygraph.o
      chgnet/graph/cygraph.c:1238:10: fatal error: 'fast_converter_libraries/create_graph.c' file not found
      #include "fast_converter_libraries/create_graph.c"
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      1 error generated.
      error: command '/usr/bin/clang' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for chgnet
  Building wheel for nvidia-ml-py3 (setup.py) ... done
  Created wheel for nvidia-ml-py3: filename=nvidia_ml_py3-7.352.0-py3-none-any.whl size=19172 sha256=390a142b23fbf3b4250902e0cd4810b5670230854a2c459a82b8ecd5c2d2a0a2
  Stored in directory: /Users/yang/Library/Caches/pip/wheels/47/50/9e/29dc79037d74c3c1bb4a8661fb608e8674b7e4260d6a3f8f51
Successfully built nvidia-ml-py3
Failed to build chgnet
ERROR: Could not build wheels for chgnet, which is required to install pyproject.toml-based projects
```


### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -51,7 +51,7 @@ jobs:
         uses: actions/checkout@v4
 
       - name: Build wheels
-        uses: pypa/cibuildwheel@v2.16.5
+        uses: pypa/cibuildwheel@v2.18.1
         env:
           CIBW_BUILD: cp${{ matrix.python-version }}-*
           CIBW_ARCHS_MACOS: universal2

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -74,10 +74,8 @@ jobs:
       - name: Download build artifacts
         uses: actions/download-artifact@v4
         with:
-          name: artifact
           path: dist
           merge-multiple: true
-          pattern: dist-*
 
       - name: Publish to PyPi or TestPyPI
         uses: pypa/gh-action-pypi-publish@release/v1

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -4,7 +4,7 @@ default_install_hook_types: [pre-commit, commit-msg]
 
 repos:
   - repo: https://github.com/astral-sh/ruff-pre-commit
-    rev: v0.4.4
+    rev: v0.4.7
     hooks:
       - id: ruff
         args: [--fix]
@@ -23,7 +23,7 @@ repos:
       - id: trailing-whitespace
 
   - repo: https://github.com/codespell-project/codespell
-    rev: v2.2.6
+    rev: v2.3.0
     hooks:
       - id: codespell
         stages: [commit, commit-msg]
@@ -46,7 +46,7 @@ repos:
           - svelte
 
   - repo: https://github.com/pre-commit/mirrors-eslint
-    rev: v9.3.0
+    rev: v9.4.0
     hooks:
       - id: eslint
         types: [file]

--- a/chgnet/graph/graph.py
+++ b/chgnet/graph/graph.py
@@ -127,7 +127,9 @@ class Graph:
         self.undirected_edges: dict[frozenset[int], list[UndirectedEdge]] = {}
         self.undirected_edges_list: list[UndirectedEdge] = []
 
-    def add_edge(self, center_index, neighbor_index, image, distance) -> None:
+    def add_edge(
+        self, center_index, neighbor_index, image, distance, dist_tol: float = 1e-6
+    ) -> None:
         """Add an directed edge to the graph.
 
         Args:
@@ -135,6 +137,8 @@ class Graph:
             neighbor_index (int): neighbor node index
             image (np.array): the periodic cell image the neighbor is from
             distance (float): distance between center and neighbor.
+            dist_tol (float): tolerance for distance comparison between edges.
+                Default = 1e-6
         """
         # Create directed_edge (DE) index using the length of added DEs
         directed_edge_index = len(self.directed_edges_list)
@@ -173,7 +177,7 @@ class Graph:
             # different image and distance (this is possible consider periodicity)
             for undirected_edge in self.undirected_edges[tmp]:
                 if (
-                    abs(undirected_edge.info["distance"] - distance) < 1e-6
+                    abs(undirected_edge.info["distance"] - distance) < dist_tol
                     and len(undirected_edge.info["directed_edge_index"]) == 1
                 ):
                     # There is an undirected edge with similar length and only one of
@@ -286,7 +290,7 @@ class Graph:
             # if encountered exception,
             # it means after Atom_Graph creation, the UDE has only 1 DE associated
             # This exception is not encountered from the develop team's experience
-            if len(u_edge.info["directed_edge_index"]) != 2:
+            if len(u_edge.info["directed_edge_index"]) != 2:  # noqa: PLR2004
                 raise ValueError(
                     "Did not find 2 Directed_edges !!!"
                     f"undirected edge {u_edge} has:"

--- a/chgnet/model/dynamics.py
+++ b/chgnet/model/dynamics.py
@@ -262,7 +262,7 @@ class StructOptimizer:
             dict[str, Structure | TrajectoryObserver]:
                 A dictionary with 'final_structure' and 'trajectory'.
         """
-        import ase.filters as filters
+        from ase import filters
         from ase.filters import Filter
 
         valid_filter_names = [

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -76,9 +76,6 @@ ignore = [
     "PLR",     # pylint refactor
     "PLW2901", # Outer for loop variable overwritten by inner assignment target
     "PT006",   # pytest-parametrize-names-wrong-type
-    "PT011",   # pytest-raises-too-broad
-    "PT013",   # pytest-incorrect-pytest-import
-    "PT019",   # pytest-fixture-param-without-value
     "PTH",     # prefer Path to os.path
     "S108",
     "S301",    # pickle can be unsafe

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -73,7 +73,9 @@ ignore = [
     "ERA001",  # found commented out code
     "ISC001",
     "NPY002",  # TODO replace legacy np.random.seed
-    "PLR",     # pylint refactor
+    "PLR0912", # too many branches
+    "PLR0913", # too many args in function def
+    "PLR0915", # too many statements
     "PLW2901", # Outer for loop variable overwritten by inner assignment target
     "PT006",   # pytest-parametrize-names-wrong-type
     "PTH",     # prefer Path to os.path
@@ -93,7 +95,16 @@ docstring-code-format = true
 
 [tool.ruff.lint.per-file-ignores]
 "site/*" = ["INP001", "S602"]
-"tests/*" = ["ANN201", "D100", "D103", "FBT001", "FBT002", "INP001", "S101"]
+"tests/*" = [
+    "ANN201",
+    "D100",
+    "D103",
+    "FBT001",
+    "FBT002",
+    "INP001",
+    "PLR2004",
+    "S101",
+]
 # E402 Module level import not at top of file
 "examples/*" = ["E402", "I002", "INP001", "N816", "S101", "T201"]
 "chgnet/**/*" = ["T201"]

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -8,7 +8,7 @@ readme = "README.md"
 license = { text = "Modified BSD" }
 dependencies = [
     "ase>=3.23.0",
-    "cython>=0.29.26",
+    "cython>=3",
     "numpy>=1.26",
     "nvidia-ml-py3>=7.352.0",
     "pymatgen>=2023.10.11",

--- a/site/make_docs.py
+++ b/site/make_docs.py
@@ -33,7 +33,8 @@ for path in glob(f"{route}/*.md"):
 
     # remove all files with less than 20 lines
     # these correspond to mostly empty __init__.py files
-    if markdown.count("\n") < 20:
+    min_line_cnt = 20
+    if markdown.count("\n") < min_line_cnt:
         os.remove(path)
         continue
 

--- a/tests/test_converter.py
+++ b/tests/test_converter.py
@@ -1,13 +1,15 @@
 from __future__ import annotations
 
-from typing import Literal
+from typing import TYPE_CHECKING, Literal
 
 import pytest
 from pymatgen.core import Lattice, Structure
-from pytest import CaptureFixture
 
 from chgnet.graph import CrystalGraph
 from chgnet.graph.converter import CrystalGraphConverter
+
+if TYPE_CHECKING:
+    from collections.abc import Generator
 
 lattice = Lattice.cubic(4)
 species = ["Na", "Cl"]
@@ -16,7 +18,7 @@ NaCl = Structure(lattice, species, coords)
 
 
 @pytest.fixture()
-def _set_make_graph() -> None:
+def _set_make_graph() -> Generator[None, None, None]:
     # fixture to force make_graph to be None and then restore it after test
     from chgnet.graph import converter
 
@@ -63,7 +65,7 @@ def test_crystal_graph_converter_warns():
 
 @pytest.mark.parametrize("on_isolated_atoms", ["ignore", "warn", "error"])
 def test_crystal_graph_converter_forward(
-    on_isolated_atoms, capsys: CaptureFixture[str]
+    on_isolated_atoms, capsys: pytest.CaptureFixture[str]
 ):
     atom_graph_cutoff = 5
     converter = CrystalGraphConverter(
@@ -75,13 +77,15 @@ def test_crystal_graph_converter_forward(
     strained.apply_strain(5)
     graph_id = "strained"
     err_msg = (
-        f"Structure {graph_id=} has 2 isolated atom(s) with "
+        f"Structure {graph_id=} has {len(NaCl)} isolated atom(s) with "
         f"{atom_graph_cutoff=}. "
         f"CHGNet calculation will likely go wrong"
     )
 
     if on_isolated_atoms == "error":
-        with pytest.raises(ValueError) as exc_info:
+        with pytest.raises(
+            ValueError, match=f"Structure {graph_id=} has {len(NaCl)} isolated atom"
+        ) as exc_info:
             converter.forward(strained, graph_id=graph_id)
         assert err_msg in str(exc_info.value)
     else:

--- a/tests/test_encoders.py
+++ b/tests/test_encoders.py
@@ -58,10 +58,8 @@ def test_angle_encoder(num_angular: int, learnable: bool) -> None:
 
 @pytest.mark.parametrize("num_angular", [-2, 8])
 def test_angle_encoder_num_angular(num_angular: int) -> None:
-    with pytest.raises(ValueError) as exc_info:
+    with pytest.raises(ValueError, match=f"{num_angular=} must be an odd integer"):
         AngleEncoder(num_angular=num_angular)
-
-    assert f"{num_angular=} must be an odd integer" in str(exc_info.value)
 
 
 @pytest.mark.parametrize("learnable", [True, False])

--- a/tests/test_graph.py
+++ b/tests/test_graph.py
@@ -100,7 +100,7 @@ def test_directed_edge() -> None:
     info = {"image": np.zeros(3), "distance": 1}
     edge = DirectedEdge([0, 1], index=0, info=info)
     undirected = edge.make_undirected(index=0, info=info)
-    assert edge == edge
+    assert edge == edge  # noqa: PLR0124
     assert edge == undirected
     assert edge.nodes == [0, 1]
     assert edge.index == 0

--- a/tests/test_md.py
+++ b/tests/test_md.py
@@ -15,7 +15,7 @@ from numpy.testing import assert_allclose
 from pymatgen.analysis.structure_matcher import StructureMatcher
 from pymatgen.core import Structure
 from pymatgen.io.ase import AseAtomsAdaptor
-from pytest import MonkeyPatch, approx
+from pytest import MonkeyPatch, approx  # noqa: PT013
 
 from chgnet import ROOT
 from chgnet.graph import CrystalGraphConverter

--- a/tests/test_model.py
+++ b/tests/test_model.py
@@ -3,7 +3,6 @@ from __future__ import annotations
 import numpy as np
 import pytest
 from pymatgen.core import Structure
-from pytest import mark
 
 from chgnet import ROOT
 from chgnet.graph import CrystalGraphConverter
@@ -14,13 +13,13 @@ graph = CrystalGraphConverter()(structure, graph_id="test-model")
 model = CHGNet.load()
 
 
-@mark.parametrize("atom_fea_dim", [1, 64])
-@mark.parametrize("bond_fea_dim", [1, 64])
-@mark.parametrize("angle_fea_dim", [1, 64])
-@mark.parametrize("num_radial", [1, 9])
-@mark.parametrize("num_angular", [1, 9])
-@mark.parametrize("n_conv", [1, 4])
-@mark.parametrize("composition_model", ["MPtrj", "MPtrj_e", "MPF"])
+@pytest.mark.parametrize("atom_fea_dim", [1, 64])
+@pytest.mark.parametrize("bond_fea_dim", [1, 64])
+@pytest.mark.parametrize("angle_fea_dim", [1, 64])
+@pytest.mark.parametrize("num_radial", [1, 9])
+@pytest.mark.parametrize("num_angular", [1, 9])
+@pytest.mark.parametrize("n_conv", [1, 4])
+@pytest.mark.parametrize("composition_model", ["MPtrj", "MPtrj_e", "MPF"])
 def test_model(
     atom_fea_dim: int,
     bond_fea_dim: int,
@@ -118,8 +117,8 @@ def test_predict_structure() -> None:
     assert out["atom_fea"].shape == (8, 64)
 
 
-@mark.parametrize("axis", [[0, 0, 1], [1, 1, 0], [-2, 3, 1]])
-@mark.parametrize("rotation_angle", [5, 30, 45, 120])
+@pytest.mark.parametrize("axis", [[0, 0, 1], [1, 1, 0], [-2, 3, 1]])
+@pytest.mark.parametrize("rotation_angle", [5, 30, 45, 120])
 def test_predict_structure_rotated(rotation_angle: float, axis: list) -> None:
     from pymatgen.transformations.standard_transformations import RotationTransformation
 

--- a/tests/test_relaxation.py
+++ b/tests/test_relaxation.py
@@ -8,7 +8,6 @@ import pytest
 import torch
 from ase.filters import ExpCellFilter, Filter, FrechetCellFilter
 from pymatgen.core import Structure
-from pytest import approx, mark, param
 
 from chgnet.graph import CrystalGraphConverter
 from chgnet.model import CHGNet, StructOptimizer
@@ -54,19 +53,20 @@ def test_relaxation(
     assert len(traj) == 2 if algorithm == "legacy" else 4
 
     # make sure final structure is more relaxed than initial one
-    assert traj.energies[-1] == approx(-58.94209, rel=1e-4)
+    assert traj.energies[-1] == pytest.approx(-58.94209, rel=1e-4)
 
 
-no_cuda = mark.skipif(not torch.cuda.is_available(), reason="No CUDA device")
+no_cuda = pytest.mark.skipif(not torch.cuda.is_available(), reason="No CUDA device")
 # skip in macos-14 M1 CI due to OOM error (TODO investigate if
 # PYTORCH_MPS_HIGH_WATERMARK_RATIO can fix)
-no_mps = mark.skipif(
+no_mps = pytest.mark.skipif(
     not torch.backends.mps.is_available() or "CI" in os.environ, reason="No MPS device"
 )
 
 
-@mark.parametrize(
-    "use_device", ["cpu", param("cuda", marks=no_cuda), param("mps", marks=no_mps)]
+@pytest.mark.parametrize(
+    "use_device",
+    ["cpu", pytest.param("cuda", marks=no_cuda), pytest.param("mps", marks=no_mps)],
 )
 def test_structure_optimizer_passes_kwargs_to_model(use_device: str) -> None:
     relaxer = StructOptimizer(use_device=use_device)

